winlogbeat: windows-11 TestPublisherMetadata/Windows_Error_Reporting/event_metadata – github.com/elastic/beats/v7/winlogbeat/sys/wineventlog

## Flaky Test

* **Test Name:** TestPublisherMetadata/Windows_Error_Reporting/event_metadata – github.com/elastic/beats/v7/winlogbeat/sys/wineventlog
* **Link:** Link to file/line number in github.
* **Branch:** https://github.com/elastic/beats/pull/32236
* **Artifact Link:** If available, attach the generated zip artifact associated with the stack trace for this failure.
* **Notes:** Additional details about the test. e.g. theory as to failure cause

### Stack Trace

```
 === RUN   TestPublisherMetadata/Windows_Error_Reporting/event_metadata
    publisher_metadata_test.go:173: The system cannot find the file specified.
        failed to open event metadata enumerator with EvtOpenEventMetadataEnum
        github.com/elastic/beats/v7/winlogbeat/sys/wineventlog.NewEventMetadataIterator
        	C:/Users/jenkins/workspace/PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe/src/github.com/elastic/beats/winlogbeat/sys/wineventlog/publisher_metadata.go:552
        github.com/elastic/beats/v7/winlogbeat/sys/wineventlog.(*PublisherMetadata).EventMetadataIterator
        	C:/Users/jenkins/workspace/PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe/src/github.com/elastic/beats/winlogbeat/sys/wineventlog/publisher_metadata.go:157
        github.com/elastic/beats/v7/winlogbeat/sys/wineventlog.testPublisherMetadata.func1.13
        	C:/Users/jenkins/workspace/PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe/src/github.com/elastic/beats/winlogbeat/sys/wineventlog/publisher_metadata_test.go:171
        testing.tRunner
        	C:/Users/jenkins/workspace/PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe/.gvm/versions/go1.17.11.windows.amd64/src/testing/testing.go:1259
        runtime.goexit
        	C:/Users/jenkins/workspace/PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe/.gvm/versions/go1.17.11.windows.amd64/src/runtime/asm_amd64.s:1581
        --- FAIL: TestPublisherMetadata/Windows_Error_Reporting/event_metadata (0.00s)
```

### Build output

```
[2022-07-07T13:44:32.019Z] 
[2022-07-07T13:44:32.019Z] C:\Users\jenkins\workspace\PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe\src\github.com\elastic\beats\winlogbeat>mage build unitTest 
[2022-07-07T13:44:50.284Z] >> build: Building winlogbeat
[2022-07-07T13:44:58.955Z] Generated fields.yml for winlogbeat to C:\Users\jenkins\workspace\PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe\src\github.com\elastic\beats\winlogbeat\build\fields\fields.yml
[2022-07-07T13:44:58.955Z] Generated fields.yml for winlogbeat to C:\Users\jenkins\workspace\PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe\src\github.com\elastic\beats\winlogbeat\build\fields\fields.common.yml
[2022-07-07T13:44:58.955Z] Generated fields.yml for winlogbeat to C:\Users\jenkins\workspace\PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe\src\github.com\elastic\beats\winlogbeat\build\fields\fields.all.yml
[2022-07-07T13:45:01.070Z] >> go test: Unit Testing
[2022-07-07T13:45:01.070Z] exec: gotestsum --no-color -f standard-quiet --junitfile build/TEST-go-unit.xml --jsonfile build/TEST-go-unit.out.json -- -tags null oracle -covermode=atomic -coverprofile=build\TEST-go-unit.cov ./...
[2022-07-07T13:45:05.684Z] ok  	github.com/elastic/beats/v7/winlogbeat/config	0.708s	coverage: 100.0% of statements
[2022-07-07T13:45:05.952Z] ok  	github.com/elastic/beats/v7/winlogbeat/sys	0.605s	coverage: 65.7% of statements
[2022-07-07T13:45:06.222Z] ok  	github.com/elastic/beats/v7/winlogbeat/sys/winevent	0.624s	coverage: 59.5% of statements
[2022-07-07T13:45:07.167Z] ok  	github.com/elastic/beats/v7/winlogbeat/beater	0.787s	coverage: 6.9% of statements
[2022-07-07T13:45:08.554Z] ok  	github.com/elastic/beats/v7/winlogbeat	0.278s	coverage: 0.0% of statements
[2022-07-07T13:45:09.937Z] FAIL
[2022-07-07T13:45:09.937Z] FAIL	github.com/elastic/beats/v7/winlogbeat/sys/wineventlog	3.480s
[2022-07-07T13:45:10.878Z] ok  	github.com/elastic/beats/v7/winlogbeat/checkpoint	5.816s	coverage: 75.5% of statements
[2022-07-07T13:45:10.878Z] ?   	github.com/elastic/beats/v7/winlogbeat/cmd	[no test files]
[2022-07-07T13:45:11.822Z] ok  	github.com/elastic/beats/v7/winlogbeat/eventlog	6.242s	coverage: 62.6% of statements
[2022-07-07T13:45:11.822Z] ?   	github.com/elastic/beats/v7/winlogbeat/include	[no test files]
[2022-07-07T13:45:11.822Z] ?   	github.com/elastic/beats/v7/winlogbeat/module	[no test files]
[2022-07-07T13:45:11.822Z] ?   	github.com/elastic/beats/v7/winlogbeat/scripts/mage	[no test files]
[2022-07-07T13:45:12.080Z] 
[2022-07-07T13:45:12.080Z] === Skipped
[2022-07-07T13:45:12.080Z] === SKIP: winlogbeat/eventlog TestBenchmarkRead (0.00s)
[2022-07-07T13:45:12.080Z]     bench_test.go:50: -benchtest not enabled
[2022-07-07T13:45:12.080Z] 
[2022-07-07T13:45:12.080Z] === SKIP: winlogbeat/sys/wineventlog TestFormatMessage/getMessageStringFromHandle/no_metadata (0.00s)
[2022-07-07T13:45:12.080Z]     format_message_test.go:45: This currently fails under Win10. The message strings are returned even though no metadata is passed.
[2022-07-07T13:45:12.080Z]         --- SKIP: TestFormatMessage/getMessageStringFromHandle/no_metadata (0.00s)
[2022-07-07T13:45:12.080Z] 
[2022-07-07T13:45:12.080Z] === SKIP: winlogbeat/sys/wineventlog TestFormatMessage/getEventXML/no_metadata (0.00s)
[2022-07-07T13:45:12.080Z]     format_message_test.go:143: This currently fails under Win10. The event XML is returned even though no metadata is passed.
[2022-07-07T13:45:12.080Z]         --- SKIP: TestFormatMessage/getEventXML/no_metadata (0.00s)
[2022-07-07T13:45:12.080Z] 
[2022-07-07T13:45:12.080Z] === Failed
[2022-07-07T13:45:12.080Z] === FAIL: winlogbeat/sys/wineventlog TestPublisherMetadata/Windows_Error_Reporting/event_metadata (0.00s)
[2022-07-07T13:45:12.080Z]     publisher_metadata_test.go:173: The system cannot find the file specified.
[2022-07-07T13:45:12.080Z]         failed to open event metadata enumerator with EvtOpenEventMetadataEnum
[2022-07-07T13:45:12.080Z]         github.com/elastic/beats/v7/winlogbeat/sys/wineventlog.NewEventMetadataIterator
[2022-07-07T13:45:12.080Z]         	C:/Users/jenkins/workspace/PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe/src/github.com/elastic/beats/winlogbeat/sys/wineventlog/publisher_metadata.go:552
[2022-07-07T13:45:12.080Z]         github.com/elastic/beats/v7/winlogbeat/sys/wineventlog.(*PublisherMetadata).EventMetadataIterator
[2022-07-07T13:45:12.080Z]         	C:/Users/jenkins/workspace/PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe/src/github.com/elastic/beats/winlogbeat/sys/wineventlog/publisher_metadata.go:157
[2022-07-07T13:45:12.080Z]         github.com/elastic/beats/v7/winlogbeat/sys/wineventlog.testPublisherMetadata.func1.13
[2022-07-07T13:45:12.080Z]         	C:/Users/jenkins/workspace/PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe/src/github.com/elastic/beats/winlogbeat/sys/wineventlog/publisher_metadata_test.go:171
[2022-07-07T13:45:12.080Z]         testing.tRunner
[2022-07-07T13:45:12.080Z]         	C:/Users/jenkins/workspace/PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe/.gvm/versions/go1.17.11.windows.amd64/src/testing/testing.go:1259
[2022-07-07T13:45:12.080Z]         runtime.goexit
[2022-07-07T13:45:12.080Z]         	C:/Users/jenkins/workspace/PR-32236-3-110895e7-0089-4aab-8a7c-d09e035e6afe/.gvm/versions/go1.17.11.windows.amd64/src/runtime/asm_amd64.s:1581
[2022-07-07T13:45:12.080Z]         --- FAIL: TestPublisherMetadata/Windows_Error_Reporting/event_metadata (0.00s)
[2022-07-07T13:45:12.080Z] 
[2022-07-07T13:45:12.080Z] === FAIL: winlogbeat/sys/wineventlog TestPublisherMetadata/Windows_Error_Reporting (0.05s)
[2022-07-07T13:45:12.081Z]     --- FAIL: TestPublisherMetadata/Windows_Error_Reporting (0.05s)
[2022-07-07T13:45:12.081Z] 
[2022-07-07T13:45:12.081Z] === FAIL: winlogbeat/sys/wineventlog TestPublisherMetadata (0.18s)
[2022-07-07T13:45:12.081Z] 
[2022-07-07T13:45:12.081Z] === FAIL: winlogbeat/sys/wineventlog TestRenderer/application-windows-error-reporting.evtx (0.03s)
[2022-07-07T13:45:12.081Z] {"log.level":"debug","@timestamp":"2022-07-07T13:45:09.943Z","log.logger":"renderer","log.origin":{"file.name":"wineventlog/metadata_store.go","file.line":261},"message":"Obtained unique event metadata from event handle. It differed from what was listed in the publisher's metadata.","publisher":"Windows Error Reporting","empty":true,"event_id":1001,"fingerprint":13417913915764132552,"default_event_metadata":null,"event_metadata":{"EventID":1001,"Version":0,"MsgStatic":"","EventData":[{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""},{"Name":"Data","InType":"","OutType":""}]},"ecs.version":"1.6.0"}
[2022-07-07T13:45:12.081Z] {"log.level":"debug","@timestamp":"2022-07-07T13:45:09.943Z","log.logger":"renderer","log.origin":{"file.name":"wineventlog/renderer.go","file.line":379},"message":"Falling back to EvtFormatMessage for event ID 1001.","ecs.version":"1.6.0"}
[2022-07-07T13:45:12.081Z]     renderer_test.go:188: Render failed: failed to load publisher metadata for Windows Error Reporting (returning an empty metadata store): failed to open event metadata enumerator with EvtOpenEventMetadataEnum: The system cannot find the file specified.
[2022-07-07T13:45:12.081Z]     --- FAIL: TestRenderer/application-windows-error-reporting.evtx (0.03s)
[2022-07-07T13:45:12.081Z] 
[2022-07-07T13:45:12.081Z] === FAIL: winlogbeat/sys/wineventlog TestRenderer (0.50s)
[2022-07-07T13:45:12.081Z] 
[2022-07-07T13:45:12.081Z] DONE 173 tests, 3 skipped, 5 failures in 11.042s
[2022-07-07T13:45:12.081Z] Error: failed to execute go: exit status 1
script returned exit code 1
```

--- a/CHANGELOG.next.asciidoc
+++ b/CHANGELOG.next.asciidoc
@@ -81,6 +81,8 @@ https://github.com/elastic/beats/compare/v8.2.0\...main[Check the HEAD diff]
 
 - Powershell: Fix processing of parameter details. {pull}31833[31833]
 - Security: Fix processing of sidlist, access list and access mask. {pull}31833[31833]
+- Fix fatal invalid memory write on Windows 11. {issue}32469[32469] {pull}32519[32519]
+- Fix handling of event formatting when no metadata is available on Windows 11. {issue}32468[32468] {pull}32519[32519]
 
 *Functionbeat*
 

--- a/winlogbeat/Jenkinsfile.yml
+++ b/winlogbeat/Jenkinsfile.yml
@@ -44,12 +44,11 @@ stages:
         platforms:             ## override default labels in this specific stage.
             - "windows-2012-r2"
         stage: extended_win
-   # See https://github.com/elastic/beats/issues/32468
-   # windows-11:
-   #     mage: "mage build unitTest"
-   #     platforms:             ## override default labels in this specific stage.
-   #         - "windows-11"
-   #     stage: extended_win
+    windows-11:
+        mage: "mage build unitTest"
+        platforms:             ## override default labels in this specific stage.
+            - "windows-11"
+        stage: extended_win
     windows-10:
         mage: "mage build unitTest"
         platforms:             ## override default labels in this specific stage.

--- a/winlogbeat/sys/wineventlog/bookmark.go
+++ b/winlogbeat/sys/wineventlog/bookmark.go
@@ -21,9 +21,9 @@
 package wineventlog
 
 import (
+	"fmt"
 	"syscall"
 
-	"github.com/pkg/errors"
 	"golang.org/x/sys/windows"
 
 	"github.com/elastic/beats/v7/winlogbeat/sys"
@@ -42,8 +42,8 @@ func (b Bookmark) XML() (string, error) {
 	var bufferUsed uint32
 
 	err := _EvtRender(NilHandle, EvtHandle(b), EvtRenderBookmark, 0, nil, &bufferUsed, nil)
-	if err != nil && err != windows.ERROR_INSUFFICIENT_BUFFER {
-		return "", errors.Wrap(err, "failed to determine necessary buffer size for EvtRender")
+	if err != nil && err != windows.ERROR_INSUFFICIENT_BUFFER { //nolint:errorlint // Bad linter! This is always errno or nil.
+		return "", fmt.Errorf("failed to determine necessary buffer size for EvtRender: %w", err)
 	}
 
 	bb := sys.NewPooledByteBuffer()
@@ -52,7 +52,7 @@ func (b Bookmark) XML() (string, error) {
 
 	err = _EvtRender(NilHandle, EvtHandle(b), EvtRenderBookmark, uint32(bb.Len()), bb.PtrAt(0), &bufferUsed, nil)
 	if err != nil {
-		return "", errors.Wrap(err, "failed to render bookmark XML")
+		return "", fmt.Errorf("failed to render bookmark XML: %w", err)
 	}
 
 	return sys.UTF16BytesToString(bb.Bytes())

--- a/winlogbeat/sys/wineventlog/format_message.go
+++ b/winlogbeat/sys/wineventlog/format_message.go
@@ -87,9 +87,13 @@ func evtFormatMessage(metadataHandle EvtHandle, eventHandle EvtHandle, messageID
 	// Get a buffer from the pool and adjust its length.
 	bb := sys.NewPooledByteBuffer()
 	defer bb.Free()
+	// The documentation for EventFormatMessage specifies that the buffer is
+	// requested "in characters", and the buffer itself is LPWSTR, meaning the
+	// characters are WCHAR so double the value.
+	// https://docs.microsoft.com/en-us/windows/win32/api/winevt/nf-winevt-evtformatmessage
 	bb.Reserve(int(bufferUsed * 2))
 
-	err = _EvtFormatMessage(metadataHandle, eventHandle, messageID, valuesCount, valuesPtr, messageFlag, uint32(bb.Len()), bb.PtrAt(0), &bufferUsed)
+	err = _EvtFormatMessage(metadataHandle, eventHandle, messageID, valuesCount, valuesPtr, messageFlag, bufferUsed, bb.PtrAt(0), &bufferUsed)
 	switch err { //nolint:errorlint // This is an errno or nil.
 	case nil: // OK
 

--- a/winlogbeat/sys/wineventlog/iterator.go
+++ b/winlogbeat/sys/wineventlog/iterator.go
@@ -21,9 +21,10 @@
 package wineventlog
 
 import (
+	"errors"
+	"fmt"
 	"sync"
 
-	"github.com/pkg/errors"
 	"golang.org/x/sys/windows"
 )
 
@@ -148,7 +149,7 @@ func (itr *EventIterator) moreHandles() bool {
 		var numReturned uint32
 
 		err := itr.evtNext(itr.subscription, batchSize, &itr.handles[0], 0, 0, &numReturned)
-		switch err {
+		switch err { //nolint:errorlint // Bad linter! This is always errno or nil.
 		case nil:
 			itr.lastErr = nil
 			itr.active = itr.handles[:numReturned]
@@ -159,8 +160,7 @@ func (itr *EventIterator) moreHandles() bool {
 				itr.subscription.Close()
 				itr.subscription, err = itr.subscriptionFactory()
 				if err != nil {
-					itr.lastErr = errors.Wrap(err, "failed in EvtNext while trying to "+
-						"recover from RPC_S_INVALID_BOUND error")
+					itr.lastErr = fmt.Errorf("failed in EvtNext while trying to recover from RPC_S_INVALID_BOUND error: %w", err)
 					return false
 				}
 
@@ -168,9 +168,7 @@ func (itr *EventIterator) moreHandles() bool {
 				batchSize = batchSize / 2
 				continue
 			} else {
-				itr.lastErr = errors.Wrap(err, "failed in EvtNext (try "+
-					"reducing the batch size or providing a subscription "+
-					"factory for automatic recovery)")
+				itr.lastErr = fmt.Errorf("failed in EvtNext (try reducing the batch size or providing a subscription factory for automatic recovery): %w", err)
 			}
 		default:
 			itr.lastErr = err

--- a/winlogbeat/sys/wineventlog/iterator_test.go
+++ b/winlogbeat/sys/wineventlog/iterator_test.go
@@ -32,7 +32,7 @@ import (
 )
 
 func TestEventIterator(t *testing.T) {
-	logp.TestingSetup()
+	logp.TestingSetup() //nolint:errcheck // Not needed.
 
 	writer, tearDown := createLog(t)
 	defer tearDown()

--- a/winlogbeat/sys/wineventlog/metadata_store.go
+++ b/winlogbeat/sys/wineventlog/metadata_store.go
@@ -21,12 +21,12 @@
 package wineventlog
 
 import (
+	"fmt"
 	"strconv"
 	"strings"
 	"sync"
 	"text/template"
 
-	"github.com/pkg/errors"
 	"go.uber.org/multierr"
 
 	"github.com/elastic/beats/v7/winlogbeat/sys"
@@ -299,7 +299,7 @@ func newEventMetadataFromEventHandle(publisher *PublisherMetadata, eventHandle E
 	// publisher metadata is unavailable or is out of sync with the events.
 	event, err := winevent.UnmarshalXML([]byte(xml))
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to unmarshal XML")
+		return nil, fmt.Errorf("failed to unmarshal XML: %w", err)
 	}
 
 	em := &EventMetadata{
@@ -404,8 +404,7 @@ func (em *EventMetadata) initEventMessage(itr *EventMetadataIterator, publisher 
 
 	msg, err := getMessageString(publisher, NilHandle, messageID, templateInserts.Slice())
 	if err != nil {
-		return errors.Wrapf(err, "failed to get message string using message "+
-			"ID %v for for event ID %v", messageID, em.EventID)
+		return fmt.Errorf("failed to get message string using message ID %v for for event ID %v: %w", messageID, em.EventID, err)
 	}
 
 	return em.setMessage(msg)
@@ -419,8 +418,7 @@ func (em *EventMetadata) setMessage(msg string) error {
 		Delims(leftTemplateDelim, rightTemplateDelim).
 		Funcs(eventMessageTemplateFuncs).Parse(msg)
 	if err != nil {
-		return errors.Wrapf(err, "failed to parse message template for "+
-			"event ID %v (template='%v')", em.EventID, msg)
+		return fmt.Errorf("failed to parse message template for event ID %v (template='%v'): %w", em.EventID, msg, err)
 	}
 
 	// One node means there were no parameters so this will optimize that case

--- a/winlogbeat/sys/wineventlog/metadata_store_test.go
+++ b/winlogbeat/sys/wineventlog/metadata_store_test.go
@@ -29,7 +29,7 @@ import (
 )
 
 func TestPublisherMetadataStore(t *testing.T) {
-	logp.TestingSetup()
+	logp.TestingSetup() //nolint:errcheck // Not needed.
 
 	s, err := NewPublisherMetadataStore(
 		NilHandle,

--- a/winlogbeat/sys/wineventlog/publisher_metadata.go
+++ b/winlogbeat/sys/wineventlog/publisher_metadata.go
@@ -21,10 +21,10 @@
 package wineventlog
 
 import (
+	"fmt"
 	"os"
 	"syscall"
 
-	"github.com/pkg/errors"
 	"go.uber.org/multierr"
 	"golang.org/x/sys/windows"
 )
@@ -59,7 +59,7 @@ func NewPublisherMetadata(session EvtHandle, name string) (*PublisherMetadata, e
 
 	handle, err := _EvtOpenPublisherMetadata(session, publisherName, logFile, 0, 0)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed in EvtOpenPublisherMetadata")
+		return nil, fmt.Errorf("failed in EvtOpenPublisherMetadata: %w", err)
 	}
 
 	return &PublisherMetadata{
@@ -79,7 +79,7 @@ func (m *PublisherMetadata) stringProperty(propertyID EvtPublisherMetadataProper
 	case nil:
 		return "", nil
 	default:
-		return "", errors.Errorf("unexpected data type: %T", v)
+		return "", fmt.Errorf("unexpected data type: %T", v)
 	}
 }
 
@@ -94,7 +94,7 @@ func (m *PublisherMetadata) PublisherGUID() (windows.GUID, error) {
 	case nil:
 		return windows.GUID{}, nil
 	default:
-		return windows.GUID{}, errors.Errorf("unexpected data type: %T", v)
+		return windows.GUID{}, fmt.Errorf("unexpected data type: %T", v)
 	}
 }
 
@@ -172,20 +172,20 @@ func NewMetadataKeywords(publisherMetadataHandle EvtHandle) ([]MetadataKeyword, 
 
 	arrayHandle, ok := v.(EvtObjectArrayPropertyHandle)
 	if !ok {
-		return nil, errors.Errorf("unexpected handle type: %T", v)
+		return nil, fmt.Errorf("unexpected handle type: %T", v)
 	}
 	defer arrayHandle.Close()
 
 	arrayLen, err := EvtGetObjectArraySize(arrayHandle)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to get keyword array length")
+		return nil, fmt.Errorf("failed to get keyword array length: %w", err)
 	}
 
 	var values []MetadataKeyword
 	for i := uint32(0); i < arrayLen; i++ {
 		md, err := NewMetadataKeyword(publisherMetadataHandle, arrayHandle, i)
 		if err != nil {
-			return nil, errors.Wrapf(err, "failed to get keyword at array index %v", i)
+			return nil, fmt.Errorf("failed to get keyword at array index %v: %w", i, err)
 		}
 
 		values = append(values, *md)
@@ -245,20 +245,20 @@ func NewMetadataOpcodes(publisherMetadataHandle EvtHandle) ([]MetadataOpcode, er
 
 	arrayHandle, ok := v.(EvtObjectArrayPropertyHandle)
 	if !ok {
-		return nil, errors.Errorf("unexpected handle type: %T", v)
+		return nil, fmt.Errorf("unexpected handle type: %T", v)
 	}
 	defer arrayHandle.Close()
 
 	arrayLen, err := EvtGetObjectArraySize(arrayHandle)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to get opcode array length")
+		return nil, fmt.Errorf("failed to get opcode array length: %w", err)
 	}
 
 	var values []MetadataOpcode
 	for i := uint32(0); i < arrayLen; i++ {
 		md, err := NewMetadataOpcode(publisherMetadataHandle, arrayHandle, i)
 		if err != nil {
-			return nil, errors.Wrapf(err, "failed to get opcode at array index %v", i)
+			return nil, fmt.Errorf("failed to get opcode at array index %v: %w", i, err)
 		}
 
 		values = append(values, *md)
@@ -318,20 +318,20 @@ func NewMetadataLevels(publisherMetadataHandle EvtHandle) ([]MetadataLevel, erro
 
 	arrayHandle, ok := v.(EvtObjectArrayPropertyHandle)
 	if !ok {
-		return nil, errors.Errorf("unexpected handle type: %T", v)
+		return nil, fmt.Errorf("unexpected handle type: %T", v)
 	}
 	defer arrayHandle.Close()
 
 	arrayLen, err := EvtGetObjectArraySize(arrayHandle)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to get level array length")
+		return nil, fmt.Errorf("failed to get level array length: %w", err)
 	}
 
 	var values []MetadataLevel
 	for i := uint32(0); i < arrayLen; i++ {
 		md, err := NewMetadataLevel(publisherMetadataHandle, arrayHandle, i)
 		if err != nil {
-			return nil, errors.Wrapf(err, "failed to get level at array index %v", i)
+			return nil, fmt.Errorf("failed to get level at array index %v: %w", i, err)
 		}
 
 		values = append(values, *md)
@@ -392,20 +392,20 @@ func NewMetadataTasks(publisherMetadataHandle EvtHandle) ([]MetadataTask, error)
 
 	arrayHandle, ok := v.(EvtObjectArrayPropertyHandle)
 	if !ok {
-		return nil, errors.Errorf("unexpected handle type: %T", v)
+		return nil, fmt.Errorf("unexpected handle type: %T", v)
 	}
 	defer arrayHandle.Close()
 
 	arrayLen, err := EvtGetObjectArraySize(arrayHandle)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to get task array length")
+		return nil, fmt.Errorf("failed to get task array length: %w", err)
 	}
 
 	var values []MetadataTask
 	for i := uint32(0); i < arrayLen; i++ {
 		md, err := NewMetadataTask(publisherMetadataHandle, arrayHandle, i)
 		if err != nil {
-			return nil, errors.Wrapf(err, "failed to get task at array index %v", i)
+			return nil, fmt.Errorf("failed to get task at array index %v: %w", i, err)
 		}
 
 		values = append(values, *md)
@@ -473,20 +473,20 @@ func NewMetadataChannels(publisherMetadataHandle EvtHandle) ([]MetadataChannel, 
 
 	arrayHandle, ok := v.(EvtObjectArrayPropertyHandle)
 	if !ok {
-		return nil, errors.Errorf("unexpected handle type: %T", v)
+		return nil, fmt.Errorf("unexpected handle type: %T", v)
 	}
 	defer arrayHandle.Close()
 
 	arrayLen, err := EvtGetObjectArraySize(arrayHandle)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to get task array length")
+		return nil, fmt.Errorf("failed to get task array length: %w", err)
 	}
 
 	var values []MetadataChannel
 	for i := uint32(0); i < arrayLen; i++ {
 		md, err := NewMetadataChannel(publisherMetadataHandle, arrayHandle, i)
 		if err != nil {
-			return nil, errors.Wrapf(err, "failed to get task at array index %v", i)
+			return nil, fmt.Errorf("failed to get task at array index %v: %w", i, err)
 		}
 
 		values = append(values, *md)
@@ -549,7 +549,7 @@ type EventMetadataIterator struct {
 func NewEventMetadataIterator(publisher *PublisherMetadata) (*EventMetadataIterator, error) {
 	eventMetadataEnumHandle, err := _EvtOpenEventMetadataEnum(publisher.Handle, 0)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to open event metadata enumerator with EvtOpenEventMetadataEnum")
+		return nil, fmt.Errorf("failed to open event metadata enumerator with EvtOpenEventMetadataEnum: %w (%#[1]v)", err)
 	}
 
 	return &EventMetadataIterator{
@@ -575,8 +575,8 @@ func (itr *EventMetadataIterator) Next() bool {
 	var err error
 	itr.currentEvent, err = _EvtNextEventMetadata(itr.eventMetadataEnumHandle, 0)
 	if err != nil {
-		if err != windows.ERROR_NO_MORE_ITEMS {
-			itr.lastErr = errors.Wrap(err, "failed advancing to next event metadata handle")
+		if err != windows.ERROR_NO_MORE_ITEMS { //nolint:errorlint // Bad linter! This is always errno or nil.
+			itr.lastErr = fmt.Errorf("failed advancing to next event metadata handle: %w", err)
 		}
 		return false
 	}
@@ -589,7 +589,7 @@ func (itr *EventMetadataIterator) Err() error {
 }
 
 func typeCastError(expected, got interface{}) error {
-	return errors.Errorf("wrong type for property. expected:%T got:%T", expected, got)
+	return fmt.Errorf("wrong type for property. expected:%T got:%T", expected, got)
 }
 
 func (itr *EventMetadataIterator) uint32Property(propertyID EvtEventMetadataPropertyID) (uint32, error) {

--- a/winlogbeat/sys/wineventlog/publisher_metadata.go
+++ b/winlogbeat/sys/wineventlog/publisher_metadata.go
@@ -548,7 +548,7 @@ type EventMetadataIterator struct {
 
 func NewEventMetadataIterator(publisher *PublisherMetadata) (*EventMetadataIterator, error) {
 	eventMetadataEnumHandle, err := _EvtOpenEventMetadataEnum(publisher.Handle, 0)
-	if err != nil {
+	if err != nil && err != windows.ERROR_FILE_NOT_FOUND { //nolint:errorlint // Bad linter! This is always errno or nil.
 		return nil, fmt.Errorf("failed to open event metadata enumerator with EvtOpenEventMetadataEnum: %w (%#[1]v)", err)
 	}
 
@@ -569,6 +569,10 @@ func (itr *EventMetadataIterator) Close() error {
 // no more items or an error occurred. You should call Err() to check for an
 // error.
 func (itr *EventMetadataIterator) Next() bool {
+	if itr.eventMetadataEnumHandle == 0 {
+		// This is only the case when we could not find the event metadata file.
+		return false
+	}
 	// Close existing handle.
 	itr.currentEvent.Close()
 

--- a/winlogbeat/sys/wineventlog/renderer.go
+++ b/winlogbeat/sys/wineventlog/renderer.go
@@ -186,7 +186,6 @@ func (r *Renderer) renderSystem(handle EvtHandle, event *winevent.Event) error {
 			continue
 		}
 
-		//nolint:errcheck // Bad linter!
 		switch property {
 		case EvtSystemProviderName:
 			event.Provider.Name = data.(string)

--- a/winlogbeat/sys/wineventlog/syscall_windows.go
+++ b/winlogbeat/sys/wineventlog/syscall_windows.go
@@ -23,7 +23,6 @@ import (
 	"time"
 	"unsafe"
 
-	"github.com/pkg/errors"
 	"golang.org/x/sys/windows"
 
 	"github.com/elastic/beats/v7/winlogbeat/sys"
@@ -237,7 +236,7 @@ const (
 	EVENTLOG_VERBOSE_LEVEL
 )
 
-// Mapping of event levels to their string representations.
+// EventLevelToString maps event levels to their string representations.
 var EventLevelToString = map[EventLevel]string{
 	EVENTLOG_LOGALWAYS_LEVEL:   "Information",
 	EVENTLOG_INFORMATION_LEVEL: "Information",
@@ -492,7 +491,7 @@ func (v EvtVariant) Data(buf []byte) (interface{}, error) {
 	case EvtVarTypeEvtHandle:
 		return EvtHandle(v.ValueAsUintPtr()), nil
 	default:
-		return nil, errors.Errorf("unhandled type: %d", typ)
+		return nil, fmt.Errorf("unhandled type: %d", typ)
 	}
 }
 
@@ -547,15 +546,15 @@ const (
 func EvtGetPublisherMetadataProperty(publisherMetadataHandle EvtHandle, propertyID EvtPublisherMetadataPropertyID) (interface{}, error) {
 	var bufferUsed uint32
 	err := _EvtGetPublisherMetadataProperty(publisherMetadataHandle, propertyID, 0, 0, nil, &bufferUsed)
-	if err != windows.ERROR_INSUFFICIENT_BUFFER {
-		return "", errors.Errorf("expected ERROR_INSUFFICIENT_BUFFER but got %v", err)
+	if err != windows.ERROR_INSUFFICIENT_BUFFER { //nolint:errorlint // Bad linter! This is always errno or nil.
+		return "", fmt.Errorf("expected ERROR_INSUFFICIENT_BUFFER but got %w (%#[1]v)", err)
 	}
 
 	buf := make([]byte, bufferUsed)
 	pEvtVariant := (*EvtVariant)(unsafe.Pointer(&buf[0]))
 	err = _EvtGetPublisherMetadataProperty(publisherMetadataHandle, propertyID, 0, uint32(len(buf)), pEvtVariant, &bufferUsed)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed in EvtGetPublisherMetadataProperty")
+		return nil, fmt.Errorf("failed in EvtGetPublisherMetadataProperty: %w", err)
 	}
 
 	v, err := pEvtVariant.Data(buf)
@@ -574,20 +573,20 @@ func EvtGetPublisherMetadataProperty(publisherMetadataHandle EvtHandle, property
 func EvtGetObjectArrayProperty(arrayHandle EvtObjectArrayPropertyHandle, propertyID EvtPublisherMetadataPropertyID, index uint32) (interface{}, error) {
 	var bufferUsed uint32
 	err := _EvtGetObjectArrayProperty(arrayHandle, propertyID, index, 0, 0, nil, &bufferUsed)
-	if err != windows.ERROR_INSUFFICIENT_BUFFER {
-		return nil, errors.Wrap(err, "failed in EvtGetObjectArrayProperty, expected ERROR_INSUFFICIENT_BUFFER")
+	if err != windows.ERROR_INSUFFICIENT_BUFFER { //nolint:errorlint // Bad linter! This is always errno or nil.
+		return nil, fmt.Errorf("failed in EvtGetObjectArrayProperty, expected ERROR_INSUFFICIENT_BUFFER: %w", err)
 	}
 
 	buf := make([]byte, bufferUsed)
 	pEvtVariant := (*EvtVariant)(unsafe.Pointer(&buf[0]))
 	err = _EvtGetObjectArrayProperty(arrayHandle, propertyID, index, 0, uint32(len(buf)), pEvtVariant, &bufferUsed)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed in EvtGetObjectArrayProperty")
+		return nil, fmt.Errorf("failed in EvtGetObjectArrayProperty: %w", err)
 	}
 
 	value, err := pEvtVariant.Data(buf)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to read EVT_VARIANT value")
+		return nil, fmt.Errorf("failed to read EVT_VARIANT value: %w", err)
 	}
 	return value, nil
 }
@@ -609,15 +608,15 @@ func EvtGetObjectArraySize(handle EvtObjectArrayPropertyHandle) (uint32, error) 
 func GetEventMetadataProperty(metadataHandle EvtHandle, propertyID EvtEventMetadataPropertyID) (interface{}, error) {
 	var bufferUsed uint32
 	err := _EvtGetEventMetadataProperty(metadataHandle, 8, 0, 0, nil, &bufferUsed)
-	if err != windows.ERROR_INSUFFICIENT_BUFFER {
-		return nil, errors.Errorf("expected ERROR_INSUFFICIENT_BUFFER but got %v", err)
+	if err != windows.ERROR_INSUFFICIENT_BUFFER { //nolint:errorlint // Bad linter! This is always errno or nil.
+		return nil, fmt.Errorf("expected ERROR_INSUFFICIENT_BUFFER but got %w (%#[1]v)", err)
 	}
 
 	buf := make([]byte, bufferUsed)
 	pEvtVariant := (*EvtVariant)(unsafe.Pointer(&buf[0]))
 	err = _EvtGetEventMetadataProperty(metadataHandle, propertyID, 0, uint32(len(buf)), pEvtVariant, &bufferUsed)
 	if err != nil {
-		return nil, errors.Wrap(err, "_EvtGetEventMetadataProperty")
+		return nil, fmt.Errorf("_EvtGetEventMetadataProperty: %w", err)
 	}
 
 	return pEvtVariant.Data(buf)

--- a/winlogbeat/sys/wineventlog/template.go
+++ b/winlogbeat/sys/wineventlog/template.go
@@ -17,9 +17,7 @@
 
 package wineventlog
 
-import (
-	"encoding/xml"
-)
+import "encoding/xml"
 
 type eventTemplate struct {
 	Data []EventData `xml:"data"`

--- a/winlogbeat/sys/wineventlog/wineventlog_windows.go
+++ b/winlogbeat/sys/wineventlog/wineventlog_windows.go
@@ -559,7 +559,6 @@ func renderXML(eventHandle EvtHandle, flag EvtRenderFlag, renderBuf []byte, out 
 	}
 
 	if int(bufferUsed) > len(renderBuf) {
-		//nolint:stylecheck // These are proper nouns.
 		return fmt.Errorf("Windows EvtRender reported that wrote %d bytes "+
 			"to the buffer, but the buffer can only hold %d bytes",
 			bufferUsed, len(renderBuf))

--- a/x-pack/winlogbeat/Jenkinsfile.yml
+++ b/x-pack/winlogbeat/Jenkinsfile.yml
@@ -49,12 +49,11 @@ stages:
         platforms:             ## override default labels in this specific stage.
             - "windows-2012-r2"
         stage: extended_win
-   # See https://github.com/elastic/beats/issues/32469
-   # windows-11:
-   #     mage: "mage build unitTest"
-   #     platforms:             ## override default labels in this specific stage.
-   #         - "windows-11"
-   #     stage: extended_win
+    windows-11:
+        mage: "mage build unitTest"
+        platforms:             ## override default labels in this specific stage.
+            - "windows-11"
+        stage: extended_win
     windows-10:
         mage: "mage build unitTest"
         platforms:             ## override default labels in this specific stage.

